subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/graph/conflicts/DefaultCapabilitiesConflictHandler.java

```diff
@@ -206,13 +206,26 @@ public void evict() {
                             public void select() {
                                 selected = node;
                             }
+
+                            @Override
+                            public void reject() {
+                                ComponentState component = node.getComponent();
+                                component.rejectForCapabilityConflict(Pair.of(capability, conflictedNodes(node, conflict.nodes)));
```

Comment:
:x: I'm not a big fan of using a `Pair` here. You could just pass the 2 arguments, and let `ComponentState` store the information as it wants. If in addition you move the error building to it, then `Pair` would never leak.

Reply:
Makes sense, refactoring